spec/condition_spec.rb

```diff
@@ -135,14 +135,333 @@
+    condition = '["not", {"name": "browser_type", "type": "custom_attribute", "value": "chrome"}]'
     condition = JSON.parse(condition)
     expect(@condition_evaluator.evaluate(condition)).to be true
   end
+
+  it 'should return nil when condition has an invalid type property' do
+    condition = '["and", {"match": "exact", "name": "weird_condition", "type": "weird", "value": "test"}]'
+    condition = JSON.parse(condition)
+    expect(@condition_evaluator.evaluate(condition)).to eq(nil)
+  end
+
+  it 'should return null when condition has an invalid match property' do
+    condition = '["and", {"match": "weird", "name": "browser_type", "type": "custom_attribute", "chrome": "test"}]'
+    condition = JSON.parse(condition)
+    expect(@condition_evaluator.evaluate(condition)).to eq(nil)
+  end
+
+  describe 'null handling' do
+    before(:context) do
+      @exact_browser_condition = {'name' => 'browser_type', 'match' => 'exact', 'type' => 'custom_attribute', 'value' => 'firefox'}
+      @exact_device_condition =  {'name' => 'device', 'match' => 'exact', 'type' => 'custom_attribute', 'value' => 'iphone'}
+      @exact_location_condition = {'name' => 'location', 'match' => 'exact', 'type' => 'custom_attribute', 'value' => 'san francisco'}
+    end
+    describe 'and evaluation' do
+      it 'should return null when all operands evaluate to null' do
+        user_attributes = {
+          'browser_type' => 4.5,
+          'location' => false
+        }
+        condition_evaluator = Optimizely::ConditionEvaluator.new(user_attributes)
+        expect(condition_evaluator.evaluate(['and', @exact_browser_condition, @exact_location_condition])).to eq(nil)
+      end
+
+      it 'should return null when operands evaluate to trues and nulls' do
+        user_attributes = {
+          'browser_type' => 'firefox',
+          'location' => false
+        }
+        condition_evaluator = Optimizely::ConditionEvaluator.new(user_attributes)
+        expect(condition_evaluator.evaluate(['and', @exact_browser_condition, @exact_location_condition])).to eq(nil)
+      end
+
+      it 'should return false when operands evaluate to falses and nulls' do
+        user_attributes = {
+          'browser_type' => 'chrome',
+          'location' => false
+        }
+        condition_evaluator = Optimizely::ConditionEvaluator.new(user_attributes)
+        expect(condition_evaluator.evaluate(['and', @exact_browser_condition, @exact_location_condition])).to be false
+      end
+
+      it 'should return false when operands evaluate to trues, falses, and nulls' do
+        user_attributes = {
+          'browser_type' => 'firefox',
+          'device' => 'android',
+          'location' => false
+        }
+        condition_evaluator = Optimizely::ConditionEvaluator.new(user_attributes)
+        expect(condition_evaluator.evaluate(['and', @exact_browser_condition, @exact_device_condition, @exact_location_condition])).to be false
```

Comment:
Let's place conditions in a way that 2nd condition evaluates to null and 3rd evaluates to False. Does it still return False?